### ISSUE TITLE
challenge's task due dates from creator differ from due date seen by participants

### ISSUE THREAD

**Issue by kawunngg** [HabitRPG/habitica#6209]
hello all,

i run a weekly challenge and participants repeatedly reported me, that they had a "wrong" due date for the challenge tasks. 

due date of the challenge should be every friday, participants often told me they saw thursday as the due date. not the same persons, not the same weeks, always thursday instead of friday.

but i did not set a wrong due date, i always see the intended due date in my list. 

i clone the challenge every week before closing it, so maybe i kind of clone the error with it?

my User-ID: ce935e7f-00e6-44bc-a44a-5280553810f0
guild: 59f699a3-9d22-4d6d-b34a-8fcdf2ad1dac
challenge URL: https://habitica.com/#/options/groups/challenges/bdc02534-6084-4c50-a861-592ac52cdd5e

User who reported wrong due date this week: 637f47a1-78e0-49b0-97ac-3b4d69fe699b

**Comment by Alys** [comment]
For both challenge and non-challenge tasks, the tasks dates are stored in the database as UTC times (GMT+0). When a player looks at the task though the website, they see the date converted to their own timezone. 

For that challenge, the dates on the To-Dos are 2015-11-12T23:00:00.000Z. Anyone with a timezone of GMT+1 and greater will see Friday; any other timezone will produce Thursday.

We need to keep storing the dates in the database as UTC. That's best practice and avoids a host of timezone-related problems.

We could in theory change the challenge joining process so that when tasks are added to a player's list, the dates on the tasks are adjusted based on the player's own timezone, but I'm not sure that's the best approach, and it would produce inconsistent results for any player who then moved to a different timezone while the challenge was running. It would also give some players an unfair advantage since they would have longer to complete the challenge than others.

I've seen some challenge owners advertise their challenge as needing to be completed by "X date in Y timezone" which is a good workaround. It at least lets the participants understand why the dates are different for them.

@kawunngg Now that I've described what's happening, I'd be interested in your thoughts about how you'd like this to be handled.

---

_EDIT 2018-12-06 adding something that I posted in a later comment, to keep all my thoughts together:_

Unfortunately fixing this is pretty tricky because in some cases you'd want a task to be due at the exact date that the owner specifies (e.g., you must finish a To-Do by a particular date) but in other cases you might want the participant's time zone to be considered - e.g., you must do a Daily every Sunday (that part would be okay because it's controlled by the repeat settings) starting on a specific Sunday (that part would fail for some participants - the Daily might not be due on that specific Sunday although it would be on the next one).

**Comment by kawunngg** [comment]
hello Alys, 

thank you for your answer. i assumed this was a time zone thing. as long as the US people reported it i did not bother. But Rupunzel asked about it because she's from UK and only has 1 hour difference. but yes, it makes sense, her time is UTC, and mine is UTC+1. 

i see a problem too with adjusting due dates to time zones. if the hoster of a challenge does not think about that, they might pick winners when participants think they still have time to finish tasks. but i realize that this problem already exists anyway due to different day starts. 
if i was in japan and my friday was over, there would still be some big amount of "due day thursday" in other parts of the world and this would not make the challenge task change. so no visible difference that the owner might already pick a winner.

the workaround would help but what bugs me is that you need to be aware of the problem to use it. 

i'm thinking of several possible solutions, but i have no idea how much work they would cause, so i'm just throwing them in:

1) owner can set a due time too. but then participants of course should be able to see that time. 

2) owner gets a warning when closing a challenge "due day is not yet over in some part of the world". (one result is that some people seem to have more time / unexpected time for completing tasks. )

3) due times are adjusted, but tasks turn inactive after time is over. owner can only close challenge when due time is over in all time zones. (alas! i'm one of those persons who always check off things a little too late ;) )

one additional thought to the time zone changing problem when due days are adjusted: isn't it somehow possible to use the system time? i renember programs that got screwed up if you changed your system time or the possibility of cheating in games with that.

**Comment by lemoness** [comment]
Is this still an issue?

**Comment by majestic-owl448** [comment]
@lemoness : yes, this still happen, the dailies or todos can be due differently for the challenge creator and some participants depending on time zone differences

**Comment by AliNiitsu93** [comment]
I've been trying to look at this problem for a while and notice, that the challenges no longer has a date-field for submitting or accepting a challenge. 
Test:
- I've created an challenge, no date field required
- I've accepted a challenge, no date-field deadlines.
People have gone around the lack of date-field by placing descriptions on their title field.
So this issue shouldn't be an issue any more?

**Comment by Alys** [comment]
@AliNiitsu93 This is for dates in tasks within a Challenge. :) If you create a To-Do, you can set an optional due date. A Daily will have a Start Date, which by default will be the current day but which the Challenge owner can change. The dates are stored as UTC date-time stamps but with a value that puts them on the correct day in the Challenge owner's time zone. A participant might see the dates being a day earlier or later, depending on their own time zone.

Unfortunately fixing this is pretty tricky because in some cases you'd want a task to be due at the exact date that the owner specifies (e.g., you must finish a To-Do by a particular date) but in other cases you might want the participant's time zone to be considered - e.g., you must do a Daily every Sunday (that part would be okay because it's controlled by the repeat settings) starting on a specific Sunday (that part would fail for some participants - the Daily might not be due on that specific Sunday although it would be on the next one).

**Comment by veeeeeee** [comment]
Removing help wanted due to status:needing discussion

**Comment by veeeeeee** [comment]
@shanaqui Can you include this issue in your review of challenge creator issues for next week?
